File: pyoffice/oa/depart.py
from os import error
from django.http import request
from django.http.response import JsonResponse
from django.shortcuts import redirect, render
from oa.models import depart,user
from oa.form import departForm
import json
from django.db.models import Count
from django.middleware.csrf import get_token ,rotate_token

def index(request):
    phone = request.session.get('phone')
    me = user.objects.filter(phone=phone).first()
    context = {}
    context['title'] = '部门管理'
    context['me'] = me
    # request.META["CSRF_COOKIE_USED"] = True
    get_token(request)
    return render(request,'depart.html',context)


def data(request):
    data = list(depart.objects.all().values())
    top_data = list(depart.objects.filter(pid=0).values())

    #统计人数
    userinfo = user.objects.values('depart').annotate(count=Count('depart'))
    countDict = {}
    for x in userinfo:
        if not x['depart']:
            continue
        countDict[x['depart']] = x['count']
    for d in data:
        d['count'] = countDict[d['id']] if d['id'] in countDict else 0


    #树   
    data = handledata(data,0)
    res = {
        'data':data,
        'menuSelectList':top_data
    }
    return JsonResponse(res,safe=False)

def handledata(data,pid):
    lst = []
    for val in data:
        if val['pid'] == pid:
            lst.append(val)
            val['children'] = handledata(data,val['id'])
    return lst

def save(request):
    if request.method == 'POST':
        params = json.loads(request.body)
        id = params.get('id')
        if id :
            obj = depart.objects.get(pk=id)
            
            forms = departForm(params,instance=obj)
            if forms.is_valid():
                # forms.save()
                obj.name = params.get('name')
                obj.pid = params.get('pid')
                obj.intro = params.get('intro')
                obj.save()
                res = {
                    'code':1,
                    'msg' : '修改成功'
                }
                return JsonResponse(res)
            else:
                error = list(forms.errors.values())[0]
                res = {
                    'code':0,
                    'msg' : error
                }
                return JsonResponse(res)
        else:
            forms = departForm(params)
            if forms.is_valid():
                data = forms.cleaned_data
                depart.objects.create(**data)
                res = {
                    'code':1,
                    'msg' : '新增成功'
                }
                return JsonResponse(res)
            else:
                error = list(forms.errors.values())[0]
                res = {
                    'code':0,
                    'msg' : error
                }
                return JsonResponse(res)

def delete(request,id):
    depart.objects.get(pk=id).delete()
    res = {
        'code':1,
        'msg' : '删除成功' 
    }
    return JsonResponse(res)


def option(request):
    data = list(depart.objects.all().values())
    data = handleOption(data,0)
    return JsonResponse(data,safe=False)

def handleOption(data,pid=0):
    lst = []
    for val in data:
        if val['pid'] == pid:
            val['label'] = val['name']
            val['value'] = val['id']
            if handleOption(data,val['id']):
                val['children'] = handleOption(data,val['id'])
            lst.append(val)
    return lst


    

File: tpOA/app/controller/Depart.php
<?php
declare (strict_types = 1);

namespace app\controller;

use app\validate\DepartValidate;
use think\exception\ValidateException;
use think\Request;
use think\facade\Cookie;
use think\facade\Session;
use think\facade\View;

class Depart
{
    /**
     * 显示资源列表
     *
     * @return \think\Response
     */
    public function index()
    {
        $phone = Session::get('phone');
        $me = \app\model\User::where('phone',$phone)->find();
        $context = [];
        $context['title'] = '部门管理';
        $context['me'] = $me;
        $csrfToken = JWT::createjwt($me['id']);
        Cookie::set('csrftoken',$csrfToken,3600);
        return View::fetch('depart',$context);
    }

    public function data()
    {
        $departs = \app\model\Depart::select();
        $top_departs = \app\model\Depart::where('pid',0)->select();
//        $departDict = \app\model\User::fieldRaw('depart,count(depart) as ct')->group('depart')->column('count(depart)','depart');
        $departDict = \app\model\User::fieldRaw('depart,count(depart) as ct')->group('depart')->where('depart','not null')->select();
        $departDict = array_column($departDict->toArray(),'ct','depart');
        //return $departDict;
        foreach ($departs as $k => $v){
            $departs[$k]['count'] = $departDict[$v['id']]??0;
        }
        //树
        $data = $this->handledata($departs,0);
        $res = [
            'data'=>$data,
            'menuSelectList'=>$top_departs
        ];
        return json($res);
    }

    public function handledata($data,$pid)
    {
        $lst = [];
        foreach ($data as $k => $v){
            if($v['pid']==$pid){
                array_push($lst,$v);
                $v['children'] = $this->handledata($data,$v['id']);
            }

        }
        return $lst;
    }
    /**
     * 显示创建资源表单页.
     *
     * @return \think\Response
     */
    public function create()
    {
        //
    }

    /**
     * 保存新建的资源
     *
     * @param  \think\Request  $request
     * @return \think\Response
     */
    public function save(Request $request)
    {
        //
        if($request->method()=='POST'){
            $id = $request->param('id');
            try{
                validate(DepartValidate::class)->check($request->param());
            }catch (ValidateException $e){
                return json(['code'=>0,'msg'=>$e->getError()]);
            }
            if(empty($id)){
                $departObj = new \app\model\Depart();
                $departObj->name = $request->param('name');
                $departObj->pid = $request->param('pid');
                $departObj->intro = $request->param('intro');
                $departObj->save();
                return json(['code'=>1,'msg'=>'添加成功']);
            }else{
                $departObj = \app\model\Depart::find($id);
                $departObj->name = $request->param('name');
                $departObj->pid = $request->param('pid');
                $departObj->intro = $request->param('intro');
                $departObj->save();
                return json(['code'=>1,'msg'=>'修改成功']);
            }
        }
    }

    /**
     * 显示指定的资源
     *
     * @param  int  $id
     * @return \think\Response
     */
    public function read($id)
    {
        //
    }

    /**
     * 显示编辑资源表单页.
     *
     * @param  int  $id
     * @return \think\Response
     */
    public function edit($id)
    {
        //
    }

    /**
     * 保存更新的资源
     *
     * @param  \think\Request  $request
     * @param  int  $id
     * @return \think\Response
     */
    public function update(Request $request, $id)
    {
        //
    }

    /**
     * 删除指定资源
     *
     * @param  int  $id
     * @return \think\Response
     */
    public function delete($id)
    {
        $departObj = \app\model\Depart::find($id);
        $departObj->delete();
        return json(['code'=>1,'msg'=>'删除成功']);
    }

    public function option()
    {
        $departs = \app\model\Depart::select();
        $data = $this->handleOption($departs,0);
        return json($data);
    }

    public function handleOption($data,$pid)
    {
        $lst = [];
        foreach($data as $k => $v){
            if($v['pid']==$pid){
                $v['label'] = $v['name'];
                $v['value'] = $v['id'];
                if($this->handleOption($data,$v['id'])){
                    $v['children'] = $this->handleOption($data,$v['id']);
                }
                $lst[] = $v;
            }
        }
        return $lst;
    }
}
